### PR TITLE
1D lithostatic pressure profile as traction boundary condition

### PR DESCRIPTION
**Pull request by anne-glerum**
@jdannberg Some cleaning up to do, hope it's clear. I've added two input files for a stokes sphere with one open boundary (2d & 3d), but no test results yet, just so you can have a look. 

I have several other test setups, of which I also have the analytical solution for the pressure profile. Would it be possible and useful to compare the calculated profile to those in a test?

Oh and I want to remove the get_traction_boundary_indicators function from simulator_access

### CODE REVIEW

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+
+          AssertThrow(spherical_representative_point[1] >= 0.0 &&
+                      spherical_representative_point[1] <= gm->opening_angle(),
+                      ExcMessage("Longitude of representative point outside shell domain."));
+
+          // set outer radius
+          spherical_representative_point[0] = gm->outer_radius();
+
+          if (dim == 3)
+            spherical_representative_point[2] *= degrees_to_radians;
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: We should probably also check if the latitude is inside of the domain.

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+
+              AssertThrow(spherical_representative_point[2] >= gm->south_latitude() &&
+                          spherical_representative_point[2] <= gm->north_latitude(),
+                          ExcMessage("Latitude of representative point outside chunk domain."));
+            }
+
+          // set outer radius
+          spherical_representative_point[0] = gm->outer_radius();
+        }
+//      TODO There will be a LayeredChunk soon :)
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: The layered chunk logic will be the same as for the chunk, right? So you can just put them into one if-statement and save some code.

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+          const Point<dim> extents = gm->get_extents();
+
+          for (unsigned int d = 0; d < dim-1; ++d)
+            AssertThrow(representative_point[d] <= extents[d], ExcMessage("Coordinate "
+                                                                          + dealii::Utilities::int_to_string(d)
+                                                                          +"Representative point outside box domain."));
+
+          // set z of surface
+          representative_point[dim-1] = extents[dim-1];
+        }
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: Same here, you can probably put that into one if?

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+      // we do not need the viscosity
+      in0.strain_rate.resize(0);
+
+      // evaluate
+      this->get_material_model().evaluate(in0, out0);
+
+      // get the magnitude of gravity. we assume
+      // that gravity always points along the depth direction. this
+      // may not strictly be true always but is likely a good enough
+      // approximation here.
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: Maybe you could add a comment in the documentation of the plugin that this is what the plugin expects, so people will not be surprised if they try a different gravity and the model does not work any more.

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+      // get the lithostatic pressure from a linear interpolation of
+      // the calculated profile
+      traction = -get_pressure(p) * normal;
+      return traction;
+    }
+
+    template <int dim>
+    double
+    LP<dim>::
+    get_pressure (const Point<dim> &p) const
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: Maybe call it 'interpolate_pressure', since that is what the function does?

File: source/simulator/simulator_access.cc [hunk 6]
```diff
@@ -364,6 +364,24 @@ namespace aspect
...
 
+   template <int dim>
+   const std::set<types::boundary_id >
+   SimulatorAccess<dim>::get_traction_boundary_indicators () const
+   {
+     std::map< types::boundary_id, std::pair<std::string, std::string> > map_indicators
+       = simulator->parameters.prescribed_traction_boundary_indicators;
+     std::map< types::boundary_id, std::pair<std::string, std::string> >::iterator it;
+     std::map< types::boundary_id, std::pair<std::string, std::string> >::iterator begin = map_indicators.begin();
+     std::map< types::boundary_id, std::pair<std::string, std::string> >::iterator end = map_indicators.end();
```
Comment (jdannberg) [inline source/simulator/simulator_access.cc]: It will make this easier to read if you just put the two lines above into the for-loop below, i.e. 
`for(it = map_indicators.begin(); it != map_indicators.end(), ++it)`

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+      // approximation here.
+      const double density0 = out0.densities[0];
+      const double gravity0 = this->get_gravity_model().gravity_vector(in0.position[0]).norm();
+
+      // now integrate pressure downward using trapezoidal integration
+      // p'(z) = rho(p,c,T) * |g| * delta_z
+      double sum = delta_z * 0.5 * density0 * gravity0;
+
+      // we assume that the t0 temperature and composition fields at the boundary
+      // are representative throughout the model time and model domain
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: I assume using the pressure from the first time step works for the models you want to run? You could also use the `update()` function, and then integrate using the temperature/composition from the current time step. That would be more expensive, but you could make it optional. 
Or would that be something that is part of the other plugin you wanted to add?

File: include/aspect/traction_boundary_conditions/lithostatic_pressure.h [hunk 8]
```diff
@@ -0,0 +1,122 @@
+/*
+  Copyright (C) 2011, 2012 by the authors of the ASPECT code.
```
Comment (jdannberg) [inline include/aspect/traction_boundary_conditions/lithostatic_pressure.h]: And please adjust the copyright years :)

File: include/aspect/traction_boundary_conditions/lithostatic_pressure.h [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+namespace aspect
+{
+  namespace TractionBoundaryConditions
+  {
+    using namespace dealii;
+
+    /**
+     * A class that implements traction boundary conditions based on a
+     * functional description provided in the input file.
```
Comment (jdannberg) [inline include/aspect/traction_boundary_conditions/lithostatic_pressure.h]: I guess this is still the old description from the function plugin?

File: include/aspect/traction_boundary_conditions/lithostatic_pressure.h [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+     *
+     * @ingroup TractionBoundaryConditionsModels
+     */
+    template <int dim>
+    class LP : public Interface<dim>, public SimulatorAccess<dim>
```
Comment (jdannberg) [inline include/aspect/traction_boundary_conditions/lithostatic_pressure.h]: And maybe give that class a longer name (like LithostaticPressure) so that it's easier to understand what it does.

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+      const double d=1.0+i-z/delta_z;
+      Assert ((d>=0) && (d<=1), ExcInternalError());
+
+      return d*pressure[i]+(1-d)*pressure[i+1];
+    }
+
+    template <int dim>
+    Point<dim>
+    LP<dim>::
+    spherical_to_cart(const Point<dim >sphere_coord) const
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: As you already noticed, the function `cartesian_coordinates` in utilities.cc does exactly the same, so you can just use that one.

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+        }
+      AssertThrow(*(traction_bi.begin()) != numbers::invalid_boundary_id,
+                  ExcMessage("Did not find the boundary indicator for the lithostatic pressure plugin."));
+
+      // the below is adapted from adiabatic_conditions/initial_profile.cc
+      // but we use the initial temperature and composition and only calculate
+      // the pressure profile
+
+      // the spacing of the depth profile
+      delta_z = this->get_geometry_model().maximal_depth() / (n_points-1);
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: Does this work correctly even though `n_points` is an integer?

Comment (bangerth) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: Yes. `double/int` yields a `double`.

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+//              AssertThrow(spherical_representative_point[2] >= gm->minimum_latitude() &&
+//                          spherical_representative_point[2] <= gm->maximum_latitude(),
+//                          ExcMessage("Representative point outside chunk domain."));
+
+              spherical_representative_point[2] *= degrees_to_radians;
+            }
+
+          // TODO this is only correct for eccentricity zero
+          // or ask for the correct radius from user
+          spherical_representative_point[0] = gm->get_semi_major_axis_a();
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: I guess we still have to think about how to do this for the EllipsoidalChunk. Is there an easy way to compute the outer radius for a given point when using the corners/eccentricity/...? Otherwise we could also add a function to the EllipsoidalChunk that computes what we want.

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+          // that gravity always points along the depth direction. this
+          // may not strictly be true always but is likely a good enough
+          // approximation here.
+          const double density = out.densities[0];
+          const double gravity = this->get_gravity_model().gravity_vector(in.position[0]).norm();
+
+          // trapezoid integration
+          pressure[i] = sum + delta_z * 0.5 * density * gravity;
+          sum += delta_z * density * gravity;
+
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: It took me a while to figure out how the integration works. If you only want to do this once at the initialization, wouldn't it be easier to take the simpler logic from the forward Euler (instead of the trapezoidal rule) and just take more points? Or have you tried that and it didn't work?

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+      else
+        {
+          // set cartesian surface coordinates
+          in0.position[0] = representative_point;
+        }
+
+      // we need the actual temperature at t0
+      in0.temperature[0] = this->get_initial_conditions().initial_temperature(in0.position[0]);
+
+      // the previous pressure
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: I guess this is not really the previous pressure, it is also the pressure at the surface for t=t0, so it might be easier just to write on comment that says we need the actual temperature, pressure and composition at the surface for t=t0.

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+      // get the magnitude of gravity. we assume
+      // that gravity always points along the depth direction. this
+      // may not strictly be true always but is likely a good enough
+      // approximation here.
+      const double density0 = out0.densities[0];
+      const double gravity0 = this->get_gravity_model().gravity_vector(in0.position[0]).norm();
+
+      // now integrate pressure downward using trapezoidal integration
+      // p'(z) = rho(p,c,T) * |g| * delta_z
+      double sum = delta_z * 0.5 * density0 * gravity0;
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: If you want to keep the trapezoidal rule, maybe you could add a short comment that you use the sum variable to integrate the pressure (maybe call it pressure_sum or pressure_integral), and then evaluate the pressure at the midpoints between the values stored in sum. 
That also makes it clear why you need the 0.5 in the first step.

File: source/traction_boundary_conditions/lithostatic_pressure.cc [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+
+          // set z of surface
+          representative_point[dim-1] = extents[dim-1];
+        }
+      else
+        AssertThrow(false, ExcNotImplemented());
+
+      // set up the input for the density function of the material model
+      typename MaterialModel::Interface<dim>::MaterialModelInputs in0(1, n_compositional_fields);
+      typename MaterialModel::Interface<dim>::MaterialModelOutputs out0(1, n_compositional_fields);
```
Comment (jdannberg) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: This is just a suggestion, but I thought you could save some code by just reusing 1 MaterialModelInputs and 1 MaterialModelOutputs, and then you can refill it every time you need to evaluate, but you would only have to set `in.strain_rate.resize(0)` once, and you wouldn't have to generate new Inputs/Outpus objects inside of your loop over all cells.

**Comment by jdannberg** [comment]
I haven't looked at the input files yet, but I really like the new functionality. Thanks!

File: source/simulator/simulator_access.cc (hunk 6 above)
Comment (anne-glerum) [inline source/simulator/simulator_access.cc]: I thought I might just remove this function, because there could be several traction boundary indicators, but with different traction boundary conditions

File: source/traction_boundary_conditions/lithostatic_pressure.cc (hunk 7 above)
Comment (anne-glerum) [inline source/traction_boundary_conditions/lithostatic_pressure.cc]: For simple models, this plugin works fine. The second plugin also takes into account along-boundary horizontal variations in temperature/composition as well as temporal variations, for multiple boundaries. Perhaps a third plugin with functionality in between these 2 is useful, but I'd need to think about that.

File: source/traction_boundary_conditions/initial_lithostatic_pressure.cc [hunk 18]
```diff
@@ -0,0 +1,441 @@
...
+
+#include <aspect/traction_boundary_conditions/initial_lithostatic_pressure.h>
+#include <aspect/global.h>
+#include <aspect/utilities.h>
+#include <deal.II/base/std_cxx11/array.h>
+
+#include <aspect/geometry_model/sphere.h>
+#include <aspect/geometry_model/spherical_shell.h>
+#include <aspect/geometry_model/chunk.h>
+//#include "layered_chunk.h"
```
Comment (jdannberg) [inline source/traction_boundary_conditions/initial_lithostatic_pressure.cc]: I guess we have to wait until the `layered_chunk` is merged before this one can be finished...

File: tests/sphere_lithostatic_pressure_2d.prm [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+############### Parameters describing the model
+# The setup is a 2d box with edge length 2890000 in which
+# all sides except the left have free slip boundary conditions. Because
+# the temperature plays no role in this model we need not
+# bother to describe temperature boundary conditions or
+# the material parameters that pertain to the temperature.
+
+
+subsection Geometry model
+  set Model name = box
```
Comment (jdannberg) [inline tests/sphere_lithostatic_pressure_2d.prm]: Your test files are called `sphere_lithostatic_pressure_...`, but they use a box geometry. That's a bit confusing. 
Also,  it would be nice to also have a test with one of the spherical geometries because there is all that extra logic that is a bit different for the spherical geometries.

File: tests/sphere_lithostatic_pressure_2d.prm [hunk 20]
```diff
@@ -0,0 +1,152 @@
+# this is a version of the stokes' law input file
+# but with less mesh refinement
```
Comment (jdannberg) [inline tests/sphere_lithostatic_pressure_2d.prm]: Could you also update the comments in the test input files, just with a very short description what the test is for?

File: tests/sphere_lithostatic_pressure_2d.prm (hunk 19 above)
Comment (anne-glerum) [inline tests/sphere_lithostatic_pressure_2d.prm]: Yes that's weird! I'll make a box and a sphere test.

File: source/traction_boundary_conditions/initial_lithostatic_pressure.cc [hunk 21]
```diff
@@ -0,0 +1,441 @@
...
+          spherical_representative_point[0] = gm->get_semi_major_axis_a();
+        }
+      else if (const GeometryModel::Sphere<dim> *gm = dynamic_cast<const GeometryModel::Sphere<dim>*> (&this->get_geometry_model()))
+        {
+          spherical_representative_point[0] = gm->radius();
+          spherical_representative_point[1] *= degrees_to_radians;
+          spherical_representative_point[2] *= degrees_to_radians;
+        }
+      // TODO: would be nice if we could combine Box and TwoMergedBoxes, but get_extents only exist for these
+      // two plugins, not all geometry models
```
Comment (jdannberg) [inline source/traction_boundary_conditions/initial_lithostatic_pressure.cc]: What do you think of the idea of @tjhei to add public functions to all of the geometry models that check if a point is within the domain (instead of doing it here), and then you could just have one call to this function (independent of the geometry model)? That could even be a different pull request, but I think it would make a lot of sense, and would make this plugin much easier to read.

File: source/traction_boundary_conditions/initial_lithostatic_pressure.cc [hunk 22]
```diff
@@ -0,0 +1,441 @@
...
+      // where to calculate the density
+      // for spherical domains
+      if (dynamic_cast<const GeometryModel::SphericalShell<dim>*> (&this->get_geometry_model()) != 0 ||
+          dynamic_cast<const GeometryModel::Sphere<dim>*> (&this->get_geometry_model()) != 0 ||
+          dynamic_cast<const GeometryModel::Chunk<dim>*> (&this->get_geometry_model()) != 0 ||
+          /*dynamic_cast<const GeometryModel::LayeredChunk<dim>*> (&this->get_geometry_model()) != 0*/
+          dynamic_cast<const GeometryModel::EllipsoidalChunk<dim>*> (&this->get_geometry_model()) != 0 )
+        {
+          // set spherical surface coordinates
+          //in0.position[0] = spherical_to_cart(spherical_representative_point);
```
Comment (jdannberg) [inline source/traction_boundary_conditions/initial_lithostatic_pressure.cc]: If the new version (with utilities) works, you probably don't need the comment with the old version any more.

File: source/traction_boundary_conditions/initial_lithostatic_pressure.cc [hunk 23]
```diff
@@ -0,0 +1,441 @@
...
+          // for spherical domains
+          if (dynamic_cast<const GeometryModel::SphericalShell<dim>*> (&this->get_geometry_model()) != 0 ||
+              dynamic_cast<const GeometryModel::Sphere<dim>*> (&this->get_geometry_model()) != 0 ||
+              dynamic_cast<const GeometryModel::Chunk<dim>*> (&this->get_geometry_model()) != 0 ||
+              /*dynamic_cast<const GeometryModel::TwoMergedChunks<dim>*> (&this->get_geometry_model()) != 0*/
+              dynamic_cast<const GeometryModel::EllipsoidalChunk<dim>*> (&this->get_geometry_model()) != 0)
+            {
+              // decrease radius with depth increment
+              spherical_representative_point[0] -= delta_z;
+              //in.position[0] = spherical_to_cart(spherical_representative_point);
```
Comment (jdannberg) [inline source/traction_boundary_conditions/initial_lithostatic_pressure.cc]: same here

File: source/traction_boundary_conditions/initial_lithostatic_pressure.cc (hunk 21 above)
Comment (bangerth) [inline source/traction_boundary_conditions/initial_lithostatic_pressure.cc]: Excellent idea! It is cheap to test whether a point is inside if you have a domain that is described by a manifold, because you just have to evaluate the pull back. But the Triangulation doesn't know this, and so figuring out whether a point is inside is really expensive in general (O(number of cells), with a big constant).

**Comment by anne-glerum** [comment]
I think this is ready for another review. 

About the tests failing: I don't know why the astyle-tester is not happy, doc/indent doesn't change anything for me. Also, the `Details` link is broken. The other testers are complaining about the statistics file of the `particle_interpolator_from_ghost_cells` test. The order of the statistics file columns has been changed.

**Comment by tjhei** [comment]
> I don't know why the astyle-tester is not happy

My fault, it is working again now.

**Comment by gassmoeller** [comment]
This PR looks pretty finished, @anne-glerum if you check that the test results are reasonable, and update the test results with the official results this might be ready for merge.

**Comment by gassmoeller** [comment]
Looks like some test failed, please take a look.
